jersey/common/src/main/java/io/helidon/jersey/common/InvokedResourceImpl.java

```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright (c) 2019 Oracle and/or its affiliates. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.helidon.jersey.common;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.ws.rs.Path;
+import javax.ws.rs.container.ContainerRequestContext;
+
+import io.helidon.common.OptionalHelper;
+
+import org.glassfish.jersey.server.ExtendedUriInfo;
+import org.glassfish.jersey.server.model.Invocable;
+import org.glassfish.jersey.server.model.ResourceMethod;
+
+final class InvokedResourceImpl implements InvokedResource {
+    private static final ConcurrentHashMap<MethodAnnotationKey, Optional<? extends Annotation>> METHOD_ANNOTATION_CACHE =
+            new ConcurrentHashMap<>();
+    private static final ConcurrentHashMap<ClassAnnotationKey, Optional<? extends Annotation>> CLASS_ANNOTATION_CACHE =
+            new ConcurrentHashMap<>();
```

Comment:
Do this caches every shrink? Could this be an issue in long-running services?

Reply:
These cache keys are class and method combination. Is there something generated in Jersey? Otherwise this is a finite set of keys depending on number of Resource classes and methods.

Reply:
"Finite" numbers can still be very large :) But this cache's size should have a small upper bound.